lib/rust/data-structures/src/unrolled_linked_list.rs

```diff
@@ -0,0 +1,1110 @@
+    ///    occupied. The preallocation can be performed either by initializing values with their
+    ///    defaults, or by zeroing the memory.
+    /// 3. Can be initialized with zeroed memory.
+    ///
+    /// [`UnrolledLinkedList`] will never automatically shrink itself, even if completely empty.
+    /// This ensures no unnecessary allocations or deallocations occur. Emptying it and then filling
+    /// it back up to the same len should incur no calls to the allocator. If you wish to free up
+    /// unused memory, use [`Self::shrink_to_fit`] or [`Self::shrink_to`].
+    ///
+    /// # Pushing new elements during iteration
+    /// Please note that as adding new elements does not require mutable access to self, it is
+    /// possible to add new elements during iteration. However, the iterator will not yield the new
+    /// elements. If you wish to iterate over the new elements, you need to create a new iterator.
+    pub struct UnrolledLinkedList
+    [T, N, I, B][T, const N: usize, I, B][T, const N: usize, I = usize, B = prealloc::Disabled] {
+        len:        Cell<usize>,
+        capacity:   Cell<usize>,
+        first_node: InitCell<ZeroableOption<Node<T, N, B>>>,
+        _index_tp:  ZST<I>,
+    }
+}
+
+impl<T, const N: usize, I, B> Default for UnrolledLinkedList<T, N, I, B> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: Debug, const N: usize, I, B> Debug for UnrolledLinkedList<T, N, I, B> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("UnrolledLinkedList")
+            .field("len", &self.len())
+            .field("capacity", &self.capacity())
+            .field("first_node", &self.first_node)
+            .finish()
+    }
+}
+
+impl<T: Debug, const N: usize, I: Index, B> Display for UnrolledLinkedList<T, N, I, B>
+where B: AllocationBehavior<T>
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_list().entries(self.iter()).finish()
+    }
+}
+
+
+// === Node Management ===
+
+impl<T, const N: usize, I, B> UnrolledLinkedList<T, N, I, B>
+where
+    I: Index,
+    B: AllocationBehavior<T>,
+{
+    #[inline(always)]
+    fn init_first_node_if_empty(&self) {
+        self.first_node.init_if_empty(|| {
+            self.capacity.set(N);
+            AllocationBehavior::new_node()
+        });
+    }
+
+    #[inline(always)]
+    fn get_or_init_first_node(&self) -> &Node<T, N, B> {
+        self.init_first_node_if_empty();
+        self.first_node.opt_item().unwrap()
+    }
+
+    #[inline(always)]
+    fn add_tail_nodes(&self, count: usize) {
+        self.get_or_init_first_node().add_tail_nodes(count);
+        self.capacity.modify(|t| *t += N * count);
+    }
+}
+
+
+// === Public API ===
+
+impl<T, const N: usize, I, B> UnrolledLinkedList<T, N, I, B> {
+    /// Constructor.
+    pub fn new() -> Self {
+        if N == 0 {
+            panic!("UnrolledLinkedList: N must be greater than 0.");
+        }
```

Comment:
This can be checked at compile-time:

```suggestion
impl<T, const N: usize, I, B> UnrolledLinkedList<T, N, I, B> {
    const _NONZERO_N: () = assert!(N > 0, "UnrolledLinkedList: N must be greater than 0.");
    
    /// Constructor.
    pub fn new() -> Self {
```

Reply:
Oh, good point!